### ISSUE TITLE
force_remove_completion does not complete token removals

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#15799]
`nodetool removenode` supports two flags in addition to the default removenode operations:
- `status`: Status of current node removal.
- `force` : Forces completion of the pending removal.
See https://docs.datastax.com/en/cassandra-oss/3.0/cassandra/tools/toolsRemoveNode.html

1. Although we have no documentation for them (See https://opensource.docs.scylladb.com/master/operating-scylla/nodetool-commands/removenode.html, @annastuchlik we need to open an issue for that as well), we do implement them in JMX and respectively over the api: `getRemovalStatus` -> `/storage_service/removal_status` and `forceRemoveCompletion` -> `/storage_service/force_remove_completion`.

2. We have zero testing in this area.  Neither in scylla unit test, nor in dtest.

3. Despite the above, users and even our field engineers apparently do use `nodetool removenode force`, and that leads to an undesired state where apparently the remove node operation is aborted, but the owned tokens are still in token_metadata but the nodes is evicted both from gossip and raft group0, rather than the desired affect of forcing the pending removal including the node tokens (which will require a following repair, but that's okay).

4. Also, `storage_service::get_removal_status` happily returns `make_ready_future<sstring>(sstring("No token removals in process."))` without even bothering to check is there's a pending removenode operation or not.  @Deexie we need to fix that as part of integrating node_ops with task_manager.

In the short term, the safest way would be to return an error from these apis until they are implemented correctly, tested, and documented.

**Comment by bhalevy** [comment]
@kbr-scylla since you worked on that for a field-urgent fix would you care to own the issue?

**Comment by bhalevy** [comment]
Cc @vladzcloudius

**Comment by asias** [comment]
In most of the cases, it is wrong to use removenode instead of replace. In the long term, we should probably deprecate the whole removenode ops. 

The only scenario that using removenode is better is that:  A node is dead and the user indeed wants to shrink the size of the cluster at the same time.  If we have no removenode ops, we could run replace then decommission.   However, if one wants to shrink the cluster size, one most probably needs to use replace ops to make each node bigger before reduce the size of the cluster anyway. Otherwise, remaining nodes would have out of disk issue.  So it is not that bad to run replace + decommission anyway. 

The benefits of deprecating removenode ops are: 

1) no more misuse of removenode where replace should be used
2) easier operation of the cluster 
3) less confusing when to use decommission or removenode
4) less node ops to support, testing / docs / procedures.

**Comment by bhalevy** [comment]
> In most of the cases, it is wrong to use removenode instead of replace. In the long term, we should probably deprecate the whole removenode ops.
> 
> The only scenario that using removenode is better is that: A node is dead and the user indeed wants to shrink the size of the cluster at the same time. If we have no removenode ops, we could run replace then decommission. However, if one wants to shrink the cluster size, one most probably needs to use replace ops to make each node bigger before reduce the size of the cluster anyway. Otherwise, remaining nodes would have out of disk issue. So it is not that bad to run replace + decommission anyway.

Shrinking the cluster has 2 reasons: use fewer, but more fatter nodes to accommodate the required storage.
Or, in case the dataset shrinked too, one may want to deploy fewer instances with the same or even smaller footprint.

For the second scenario, using removenode still makes sense.

But, with tablets, we actually have reasons not to support replace, but rather rely on bootstrap to add new resources, and then use decommission for live nodes or removenode for dead nodes.
So in the case of a lost node, the sensible path would be to bootstrap an empty node, then remove the dead node.

We may want instead to limit the support for removenode or replace based on vnodes vs. tablets partitioning.

> 
> The benefits of deprecating removenode ops are:
> 
> 1. no more misuse of removenode where replace should be used
> 2. easier operation of the cluster
> 3. less confusing when to use decommission or removenode
> 4. less node ops to support, testing / docs / procedures.

**Comment by asias** [comment]
> > In most of the cases, it is wrong to use removenode instead of replace. In the long term, we should probably deprecate the whole removenode ops.
> > The only scenario that using removenode is better is that: A node is dead and the user indeed wants to shrink the size of the cluster at the same time. If we have no removenode ops, we could run replace then decommission. However, if one wants to shrink the cluster size, one most probably needs to use replace ops to make each node bigger before reduce the size of the cluster anyway. Otherwise, remaining nodes would have out of disk issue. So it is not that bad to run replace + decommission anyway.
> 
> Shrinking the cluster has 2 reasons: use fewer, but more fatter nodes to accommodate the required storage. Or, in case the dataset shrinked too, one may want to deploy fewer instances with the same or even smaller footprint.

> 
> For the second scenario, using removenode still makes sense.

That's exactly my point: we have a limited scenario where removenode makes sense. But it can be achieved with replace + decommission. The benefits of get rid of removenode ops is bigger in my view. 

> 
> But, with tablets, we actually have reasons not to support replace, but rather rely on bootstrap to add new resources, and then use decommission for live nodes or removenode for dead nodes. So in the case of a lost node, the sensible path would be to bootstrap an empty node, then remove the dead node.
> 
> We may want instead to limit the support for removenode or replace based on vnodes vs. tablets partitioning.
> 
> > The benefits of deprecating removenode ops are:
> > 
> > 1. no more misuse of removenode where replace should be used
> > 2. easier operation of the cluster
> > 3. less confusing when to use decommission or removenode
> > 4. less node ops to support, testing / docs / procedures.

**Comment by bhalevy** [comment]
@kbr-scylla, as a first line of defense I suggest to submit a patch to return an error from `storage_service::get_removal_status` and `storage_service::force_remove_completion` until they are fixed.
And that should be backported to all live versions.

Then, we can follow up with either fixing the api implementation or deprecating removenode as @asias suggested.
But until then, the safest coarse of action is to return an error from the implementation.

**Comment by kostja** [comment]
@bhalevy how are these labels justified?

**Comment by kostja** [comment]
@kbr-scylla one thing is certain, removenode force and removenode status needs to be tested with raft-based topology changes. I updated the test plan.

**Comment by kbr-scylla** [comment]
> Despite the above, users and even our field engineers apparently do use nodetool removenode force, and that leads to an undesired state where apparently the remove node operation is aborted, but the owned tokens are still in token_metadata but the nodes is evicted both from gossip and raft group0, rather than the desired affect of forcing the pending removal including the node tokens (which will require a following repair, but that's okay).


Did you test this hypothesis?

**Comment by kbr-scylla** [comment]
> In most of the cases, it is wrong to use removenode instead of replace. In the long term, we should probably deprecate the whole removenode ops.

I don't agree. It's simpler to get rid of a dead node this way than replace+decommission. Sometimes the priority is to get rid of a dead node ASAP, even when you don't yet have a replacement for it. Sometimes replace doesn't work and you have to `removenode` (as we saw recently with a customer case).

`removenode` has saved our butts multiple times in the field.

**Comment by bhalevy** [comment]
> @bhalevy how are these labels justified?

By me woken at f'in 2:30 am to help save the world after some admin shot himself and the cluster in the foot by running `nodetool removenode force` :)

How is it an `enhancement` if the api doesn't work as expected and causes havoc in the system?!!

**Comment by bhalevy** [comment]
And as much as it would be great to get it working with raft consistent topology changes,
we need some fix downstream - as an immediate remedy, let's quickly submit and backport a patch that returns an error from the malfunctioning api rather than doing the wrong thing.
Then later we can consider actually fixing it, if that will take more time.
Though skipping the data sync phase should be a quick enough fix anyhow, warranted it synchronizes with a running removendoe, if there is one (but it doesn't have to be IMO) and modulo testing.

**Comment by kbr-scylla** [comment]
I performed the following experiments, all on a 3 node cluster {A, B, C}, node C being dead. I injected a 10 second sleep during the streaming phase to prolong it (my clusters had no data).

First:
1. start `remove_node` on A to remove C
2. during streaming phase, run `force_remove_completion` on A

Result: `force_remove_completion` waited until the `remove_node` completed, then did nothing.

Second:
1. start `remove_node` on A to remove C
2. during streaming phase, run `force_remove_completion` on B
3. kill A before it leaves the streaming phase

Result: `force_remove_completion` immediately removed C from the cluster. The removal appears to be correct: nodetool status doesn't show the node, and gossipinfo shows it as `removed` status. On A as well once it got restarted.


Third:
Same as second, except don't kill A.
Result: same as second, and once A finished the streaming phase, it "removed" C again, which had no additional effect (the removal step is basically idempotent).

So... I see no evidence of such thing happening:
> Despite the above, users and even our field engineers apparently do use nodetool removenode force, and that leads to an undesired state where apparently the remove node operation is aborted, but the owned tokens are still in token_metadata but the nodes is evicted both from gossip and raft group0, rather than the desired affect of forcing the pending removal including the node tokens (which will require a following repair, but that's okay).

I did the experiments on `master` though, so maybe it is different on other versions, but did we even modify `force_remove_completion` at any point?

In any case, I would like to see the evidence behind above claim. In the case where we saw this weird cluster state, how do we know it was `removenode force` that caused this, and not something else? I know that the administrator of that cluster did a lot of random weird stuff to it...

At this point, I see no reason to rush getting rid of `removenode force`, it seems to work as intended, at least from the above simple experiments.

**Comment by bhalevy** [comment]
Ref https://github.com/scylladb/scylladb/pull/15834

**Comment by kostja** [comment]
@kbr-scylla either way with raft-based-topology changes we can expect things to work consistently, and removenode force just becomes an alias for removenode. So I think it would be sufficient to add the test for this.
